worlds/psychonauts/__init__.py

```diff
@@ -0,0 +1,262 @@
+        # small_count = excess_arrowheads // bundle_difference
+        # large_count = num_deep_arrowhead_locations - small_count
+
+        # Adjust how many of each item to add to the item pool.
+        item_counts[ItemName.AHSmall] += small_count
+        item_counts[ItemName.AHLarge] += large_count
+
+    @staticmethod
+    def _add_mental_cobweb_shuffle_items(item_counts: Dict[str, int]):
+        # A single Mental Cobweb can normally be turned into a PSI Card at the loom in Ford's Sanctuary, so add as many
+        # PSI Cards to the pool as Mental Cobweb Locations.
+        item_counts[ItemName.PsiCard] += len(MENTAL_COBWEB_LOCATIONS)
+
+    def create_items(self):
+        """
+        Fills ItemPool 
+        """
+        num_locations_to_fill = len(self.multiworld.get_unfilled_locations(self.player))
+
+        adjusted_item_counts = ITEM_COUNT.copy()
+
+        # Pre-collect starting minds and remove them from the item pool.
+        num_starting_minds = self.options.RandomStartingMinds.value
+        if num_starting_minds > 0:
+            mind_unlocks = list(MINDS)
+            for _ in range(num_starting_minds):
+                # Pop a random mind from the list.
+                item = mind_unlocks.pop(self.random.randrange(len(mind_unlocks)))
+                # Reduce the number to add to the pool.
+                adjusted_item_counts[item] -= 1
+                self.multiworld.push_precollected(self.create_item(item))
+
+        # Pre-collect the Cobweb Duster when starting with it.
+        if self.options.StartingCobwebDuster:
+            # Reduce the count to add to the item pool.
+            adjusted_item_counts[ItemName.CobwebDuster] -= 1
+            self.multiworld.push_precollected(self.create_item(ItemName.CobwebDuster))
+
+        # Add items for DeepArrowheadShuffle
+        if self.options.DeepArrowheadShuffle:
+            self._add_deep_arrowhead_shuffle_items(adjusted_item_counts)
+
+        # Add items for MentalCobwebShuffle
+        if self.options.MentalCobwebShuffle:
+            self._add_mental_cobweb_shuffle_items(adjusted_item_counts)
+
+        # Create the initial item pool.
+        item_pool = list(map(self.create_item, repeated_item_names_gen(ITEM_DICTIONARY, adjusted_item_counts)))
+
+        assert len(item_pool) <= num_locations_to_fill, ("The initial item pool cannot be larger than the number of"
+                                                         " unfilled locations.")
+        num_locations_to_fill -= len(item_pool)
+
+        # Add filler/junk items to fill out the remaining locations.
+        # If there are more locations remaining than the desired maximum number of filler items to add, also add the
+        # Feather and Watering Can junk items to the pool.
+        desired_max_filler = 107  # This is arbitrary based on the max number of Psi Cards placeable in the game world.
```

Comment:
Any items more than what the game can place in the game world are replaced with AP placeholder items, usually used for items belonging to other worlds. The number of AP placeholder items that can be placed in the game world is equal to the number of locations that place items into the game world, so it should not be possible to run out.

Reply:
Well that's good at least, but can the client/game handle more than 107 Psi Cards or more than 1 Watering Can or Feather, or duplicate Brains, and so on?

Reply:
Yes, the game handles this. The game splits up the items into those for which duplicates can exist (usually consumables or upgrades that increase a maximum count of a resource) and those for which duplicates cannot (usually inventory items). If the game receives multiple of an item for which duplicates cannot exist, it simply does not grant any extra copies to the player if they have already received one of that item.

For the items for which duplicates can exist, I tested sending myself silly amounts of them through the `/send_multiple` server command to make sure they wouldn't break. Most just increase a count stored as a lua integer which shouldn't get anywhere near close to the maximum integer value.

There are some items where granting extra copies would work, but is undesirable because there would be no point to having extras and most of these items are permanently added to the inventory. The inventory is not well-designed for having many items in it at one time and items cannot be freely removed from it either.